mjolnir/interaction/local/DirectionalContactInteraction.hpp

```diff
@@ -0,0 +1,390 @@
+                                const real_type            margin = 0.5)
+      : kind_(kind), potentials_(pot), margin_(margin)
+  {}
+  DirectionalContactInteraction(const connection_kind_type kind,
+                                container_type&&           pot,
+                                const real_type            margin = 0.5)
+      : kind_(kind), potentials_(pot), margin_(margin)
+  {}
+  ~DirectionalContactInteraction() override{}
+
+  void      calc_force (system_type&) const noexcept override;
+  real_type calc_energy(const system_type&) const noexcept override;
+
+  void initialize(const system_type& sys) override
+  {
+    MJOLNIR_GET_DEFAULT_LOGGER();
+    MJOLNIR_LOG_FUNCTION();
+    MJOLNIR_LOG_INFO("angle1 potential = ", angle1_pot_type::name(),
+                     ", angle2 potential = ", angle2_pot_type::name(),
+                     ", contact potential = ", contact_pot_type::name(),
+                     ", number of contacts = ", potentials_.size());
+
+    this->cutoff_ = this->max_cutoff_length();
+    this->make_list(sys);
+    return;
+  }
+
+  void update(const system_type& sys) override
+  {
+    for(auto& item: potentials_)
+    {
+      std::get<3>(item).update(sys);
+    }
+
+    this->cutoff_ = this->max_cutoff_length();
+    this->make_list(sys);
+    return;
+  }
+
+  void update_margin(const real_type dmargin, const system_type& sys) override
+  {
+    this->current_margin_ -= dmargin;
+    if(this->current_margin_ < 0)
+    {
+      this->make_list(sys);
+    }
+    return;
+  }
+
+  std::string name() const override
+  {
+    return "DirectionalContact:"_s + angle1_pot_type::name()
+        + ","_s + angle2_pot_type::name() + ","_s + contact_pot_type::name();
+  }
+
+  void write_topology(topology_type&) const override;
+  container_type const& potentials() const noexcept {return potentials_;}
+  container_type&       potentials()       noexcept {return potentials_;}
+
+ private:
+
+  void make_list(const system_type& sys)
+  {
+    this->active_contacts_.clear();
+    this->active_contacts_.reserve(potentials_.size());
+
+    // absolute length of margin (this->margin_ is a relative length).
+    const real_type abs_margin = this->cutoff_ * this->margin_;
+
+    for(std::size_t i=0;i < this->potentials_.size(); ++i)
+    {
+      const indices_potentials_tuple& pot = this->potentials_[i];
+      const coordinate_type pos0 = sys.position(std::get<0>(pot)[1]);
+      const coordinate_type pos1 = sys.position(std::get<0>(pot)[2]);
+      const coordinate_type dpos = sys.adjust_direction(pos1 - pos0);
+      const real_type       len2 = math::length_sq(dpos);
+
+      const real_type rc = std::get<3>(pot).cutoff() + abs_margin;
+      if(len2 < rc * rc)
+      {
+        this->active_contacts_.push_back(i);
+      }
+      this->current_margin_ = this->cutoff_ * this->margin_;
+    }
+    return;
+  }
+
+  const real_type max_cutoff_length() const noexcept
```

Comment:
meaningless `const`.
```suggestion
  real_type max_cutoff_length() const noexcept
```